pinot-core/src/test/java/org/apache/pinot/queries/BaseQueriesTest.java

```diff
@@ -146,14 +160,21 @@ private BrokerResponseNative getBrokerResponse(String query, PlanMaker planMaker
   /**
    * Run query on multiple index segments with custom plan maker.
    * <p>Use this to test the whole flow from server to broker.
-   * <p>The result should be equivalent to querying 4 identical index segments.
+   * <p>Unless explicitly override getDistinctInstances or initialize 2 distinct index segments in test, the result
+   * should be equivalent to querying 4 identical index segments. Otherwise, the result will be equivalent to
+   * querying 2 distinct instances each having 2 segments.
    */
   private BrokerResponseNative getBrokerResponse(PinotQuery pinotQuery, PlanMaker planMaker) {
     PinotQuery serverPinotQuery = GapfillUtils.stripGapfill(pinotQuery);
     QueryContext queryContext = QueryContextConverterUtils.getQueryContext(pinotQuery);
     QueryContext serverQueryContext =
         serverPinotQuery == pinotQuery ? queryContext : QueryContextConverterUtils.getQueryContext(serverPinotQuery);
 
+    List<List<IndexSegment>> instances = getDistinctInstances();
+    if (instances.size() == 2) {
+      return getBrokerResponseDistinctInstances(pinotQuery, planMaker);
+    }
+
```

Comment:
This appears to be hardcoding how a query will be executed when there are two segments. If I understand correctly there are two problems here: 1) the result is being duplicated which is causing problem with covariance values, and 2) ability to run queries against multiple segment.  I ran into the same issue while testing a feature that I am working on. The first problem can be solved by adding a boolean variable for not duplicating the query results. The second (executing queries against multiple segments can be solved by local setup in the `CovarianceQueriesTest.java`. I will share the code with you offline. Let's see if we can do this without hardcoding for 2 segments.

Reply:
@amrishlal  1) causes problem for all non-additive statistical values, e.g. average, standard deviation, covariance, median, percentile ... for these values the result querying 4 identical segments is no different from querying one single segment, as can be potential problematic for full functional testing. For 2) can you elaborate more? I think the code currently can run on multiple segments, but people are just setting two identical segments?

Reply:
@jasperjiaguo Yes, the code can run on multiple segments, but we are duplicating segments in many test cases and we also duplicate server in BaseQueriesTest.java to mimic realtime server. I think @SabrinaZhaozyf is adding the case for running queries with two servers on two segments (?). At some point, I guess, this file could be cleaned up a bit to setup all the common cases more cleanly: 1) single server single segment, 2) single server multiple segments, 3) two servers, one segement each, etc... But that is probably another PR.

I am ok with the if check ( `if (instances.size() == 2)`). I confused this a bit with other cases.